### PR TITLE
🐛 Fix boundary event interruption

### PR DESCRIPTION
**Pull request by ElRaptorus**
**Changes:**

1. Make `ProcessTokenFacade` run synchronously. It never made any kind of sense to have it run in Promises, because it doesn't actually perform any kind of async operations.
2. Add new `FlowNodeHandlerInterruptable` base handler, which provides functionality for interrupting a handler. This allows for the BoundaryEvents to be actually interruptible.
3. Derive all activity handlers from the new `FlowNodeHandlerInterruptable` and implement support for interruption.
4. Switch to Bluebird Promises to support Promise cancellation
    - This allows a handler to actually abort its operation, as soon as an interrupt-call is made. I tested this and can confirm that as soon as `BluebirdPromise.cancel` is called, the Bluebird-based promise is truly aborted.
    - Bluebird also provides full stack traces with rejected Promises, which is immensely useful for debugging.
    - Also, since Bluebird runs faster than native promises and is able to free up resources from canceled promises, we received a significant performance boost.
    - **NOTE**: Service Tasks will not be affected by this, since they delegate their work to external components, over which we have no control. If those use native Promises, then there is no way to cancel those.
5. Every BoundaryEvent now calls its decorated handlers `interrupt` function, as soon as the triggering event occurs.
    - BoundaryEvents also provide an `interrupt` function of their own, to ensure that chained BoundaryEvents will pass to call through to the handler.
6. Refactor the TimerBoundaryEvent handler - Timer subscription is now only done in one separate function, just as it is done in the other handlers 
7. Remove some unnecessary Promise wrapping from synchronous functions.

**Issues:** 

Implements: https://github.com/process-engine/process_engine_runtime/issues/176

PR: #218

**Note**:
Since we cannot interrupt a running ProcessInstance yet, the handlers for Subprocesses and CallActivities have not been refactored, because there is little point in doing so.

## How can others test the changes?

Run mutliple scenarios where you have interrupting boundary events and see that their decorated handlers will now truly be canceled, as soon as the interrupt-command is given.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Fixing Security Issues   | `:lock:`             | 🔒     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️      |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️      |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/error_boundary_event_handler.ts [hunk 1]
```diff
@@ -11,38 +11,41 @@ import {
...
 
-  private get errorBoundaryEvent(): Model.Events.BoundaryEvent {
-    return super.flowNode;
+  // Since ErrorBoundaryEvents can be part of a BoundaryEventChain, they must also implement this method,
+  // so they can tell their decorated handler to abort.
+  public async interrupt(token: Runtime.Types.ProcessToken, terminate?: boolean): Promise<void> {
+
+    // TODO: This check can probably be removed as soon as CallActivities and SubprocessHandlers can be interrupted.
+    // But that will not happen until we are able to interrupt Subprocesses.
+    const handlerIsInterruptible: boolean = typeof (this._decoratedHandler as any).interrupt === 'function';
```
Comment (roblen45) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/error_boundary_event_handler.ts]: Ist das notwendig? Es handelt sich doch immer um einen FlowNodeHandlerInterruptible oder?

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/signal_boundary_event_handler.ts [hunk 2]
```diff
@@ -41,34 +42,50 @@ export class SignalBoundaryEventHandler extends FlowNodeHandler<Model.Events.Bou
...
+  public async interrupt(token: Runtime.Types.ProcessToken, terminate?: boolean): Promise<void> {
+
+    if (this.subscription) {
+      this.subscription.dispose();
+    }
+    this.handlerPromise.cancel();
+
+    // TODO: This check can probably be removed as soon as CallActivities and SubprocessHandlers can be interrupted.
+    // But that will not happen until we are able to interrupt Subprocesses.
+    const handlerIsInterruptible: boolean = typeof (this._decoratedHandler as any).interrupt === 'function';
```
Comment (roblen45) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/signal_boundary_event_handler.ts]: s.o.

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/timer_boundary_event_handler.ts [hunk 3]
```diff
@@ -38,58 +44,49 @@ export class TimerBoundaryEventHandler extends FlowNodeHandler<Model.Events.Boun
...
+  public async interrupt(token: Runtime.Types.ProcessToken, terminate?: boolean): Promise<void> {
+
+    if (this.timerSubscription) {
+      this.timerSubscription.dispose();
+    }
+    this.handlerPromise.cancel();
+
+    // TODO: This check can probably be removed as soon as CallActivities and SubprocessHandlers can be interrupted.
+    // But that will not happen until we are able to interrupt Subprocesses.
+    const handlerIsInterruptible: boolean = typeof (this._decoratedHandler as any).interrupt === 'function';
```
Comment (roblen45) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/timer_boundary_event_handler.ts]: s.o.

File: src/runtime/engine/flow_node_handler/boundary_event_handlers/message_boundary_event_handler.ts [hunk 4]
```diff
@@ -41,34 +42,50 @@ export class MessageBoundaryEventHandler extends FlowNodeHandler<Model.Events.Bo
...
+  public async interrupt(token: Runtime.Types.ProcessToken, terminate?: boolean): Promise<void> {
+
+    if (this.subscription) {
+      this.subscription.dispose();
+    }
+    this.handlerPromise.cancel();
+
+    // TODO: This check can probably be removed as soon as CallActivities and SubprocessHandlers can be interrupted.
+    // But that will not happen until we are able to interrupt Subprocesses.
+    const handlerIsInterruptible: boolean = typeof (this._decoratedHandler as any).interrupt === 'function';
```
Comment (roblen45) [inline src/runtime/engine/flow_node_handler/boundary_event_handlers/message_boundary_event_handler.ts]: s.o.

**Comment by ElRaptorus** [comment]
@roblen45 done

**Pull request merged**